src/replication.c

```diff
@@ -1324,20 +1382,43 @@ check_for_command_completion(spec_t *spec, rcommon_cmd_t *rcomm_cmd, ISTGT_LU_CM
 		}
 	}
 
+	response_received = success + failure;
+	min_response = MAX(rcomm_cmd->replication_factor -
+	    rcomm_cmd->consistency_factor + 1, rcomm_cmd->consistency_factor);
+
 	if (rcomm_cmd->opcode == ZVOL_OPCODE_READ) {
-		if ((rcomm_cmd->copies_sent != (success + failure))) {
-			rc = 0;
-		} else if (rcomm_cmd->copies_sent == failure) {
-			rc = -1;
+		response_received = success + failure;
+
+		if (rcomm_cmd->copies_sent >= min_response) {
+			if (response_received < min_response) {
+				rc = 0;
+			} else {
+				if (failure >= min_response ||
+				    success < min_response) {
+					rc = -1;
+				} else {
+					data = handle_read_consistency(
+					    rcomm_cmd, spec->blocklen,
+					    true);
+					cmd->data = data;
+					rc = 1;
+				}
+			}
 		} else {
-			data = handle_read_consistency(rcomm_cmd);
-			cmd->data = data;
-			rc = 1;
+			if (rcomm_cmd->copies_sent == success) {
```

Comment:
add comment that copies_sent should be 1 in this case

Reply:
ok